Tweak windows introduction page

sftim opened the pull request:
Revise [Intro to Windows support in Kubernetes](https://kubernetes.io/docs/setup/production-environment/windows/intro-windows-in-kubernetes/) [[preview](https://deploy-preview-26848--kubernetes-io-main-staging.netlify.app/docs/setup/production-environment/windows/intro-windows-in-kubernetes/)] to align better with the [style guide](https://kubernetes.io/docs/contribute/style/style-guide/), and to incorprorate content corrections raised during PR review.

- don't refer to the reader as a “customer”
- remove several statements about the future
- reorder content and reduce heading levels
etc etc etc

/kind cleanup
/language en

celestehorgan commented:
@sftim 

Rebase please! :)

sftim commented:
(rebased)

kbhawkey commented:
/cc @kubernetes/sig-windows-misc

marosset commented:
/assign

marosset commented on content/en/docs/setup/production-environment/windows/intro-windows-in-kubernetes.md:
```diff
@@ -2,118 +2,77 @@
-
-Refer to the following table for Windows operating system support in Kubernetes. A single heterogeneous Kubernetes cluster can have both Windows and Linux worker nodes. Windows containers have to be scheduled on Windows nodes and Linux containers on Linux nodes.
-
-| Kubernetes version | Windows Server LTSC releases | Windows Server SAC releases |
```
I think we should keep the Windows OS Version Support section in there.
Users frequently ask which Windows versions have been tested to work with Kubernetes and also there is occasionally functionality added to Kubernetes that depends on OS support.

sftim replied:
This is a long page for an _introduction_ - could I move the support table to the reference section and hyperlink there?

marosset replied:
That should work.

marosset commented on content/en/docs/setup/production-environment/windows/intro-windows-in-kubernetes.md:
```diff
@@ -123,40 +82,19 @@ Docker EE-basic 19.03+ is the recommended container runtime for all Windows Serv
+Microsoft maintains a Windows pause infrastructure container at `mcr.microsoft.com/oss/kubernetes/pause:1.4.1`.
 
-{{< feature-state for_k8s_version="v1.19" state="beta" >}}
+### Container runtimes {#container-runtime}
```
I think this ### Container runtimes may be a duplicate and not needed.

marosset commented:
Overall looks good to me!
Thanks @sftim 

/cc @jsturtevant @jayunit100 (SIG-Windows TLs) in case they would also like to comment/review.

aravindhp commented on content/en/docs/setup/production-environment/windows/intro-windows-in-kubernetes.md:
```diff
@@ -231,73 +169,113 @@ On Windows, using IPv6 with Kubernetes require Windows Server, version 2004 (ker
+
+### Operating system restrictions
+
+Windows has strict compatibility rules, where the host OS version must match the container base image OS version. Only Windows containers with a container operating system of Windows Server 2019 are supported.
```
Should we link to https://docs.microsoft.com/en-us/virtualization/windowscontainers/deploy-containers/version-compatibility?tabs=windows-server-20H2%2Cwindows-10-20H2 somewhere here as it might help the user?

sftim replied:
That might be a separate update - this one's already large.

aravindhp commented on content/en/docs/setup/production-environment/windows/intro-windows-in-kubernetes.md:
```diff
@@ -318,22 +295,19 @@ The following networking functionality is not supported on Windows nodes
 
 1. Node-pod connectivity isn't possible by design. It's only possible for local pods with Flannel v0.12.0 (or higher).
-2. We are restricted to using VNI 4096 and UDP port 4789. The VNI limitation is being worked on and will be overcome in a future release (open-source flannel changes). See the official [Flannel VXLAN](https://github.com/coreos/flannel/blob/master/Documentation/backends.md#vxlan) backend docs for more details on these parameters.
+2. Flannel is restricted to using VNI 4096 and UDP port 4789. See the official [Flannel VXLAN](https://github.com/coreos/flannel/blob/master/Documentation/backends.md#vxlan) backend docs for more details on these parameters.
```
Might be useful to call out https://docs.microsoft.com/en-us/virtualization/windowscontainers/kubernetes/common-problems#pod-to-pod-connectivity-between-hosts-is-broken-on-my-kubernetes-cluster-running-on-vsphere in the context of port 4789.

sftim replied:
That might be a separate update - this one's already large.

aravindhp reviewed (commented):
Thanks for doing this @sftim. This will doc layout much better. Just a few comments.

sftim commented on content/en/docs/setup/production-environment/windows/intro-windows-in-kubernetes.md:
```diff
@@ -2,118 +2,77 @@
-{{< /note >}}
-{{< note >}}
-Windows containers with process isolation have strict compatibility rules, [where the host OS version must match the container base image OS version](https://docs.microsoft.com/en-us/virtualization/windowscontainers/deploy-containers/version-compatibility). Once we support Windows containers with Hyper-V isolation in Kubernetes, the limitation and compatibility rules will change.
+This document uses the term “Windows containers” to mean Windows containers with process isolation. Kubernetes does not support running Windows containers with [Hyper-V isolation](https://docs.microsoft.com/en-us/virtualization/windowscontainers/manage-containers/hyperv-container).
```
That might be a separate update. If the existing document is incorrect we could fix that separately then I can rebase.

I don't know if containerd and Hyper-V isolation works; it's well outside my area of technical expertise.

jsturtevant replied:
containerd with process isolation today works.  Hyper-v will be possible but is not currently up streamed in containerd and not supported in k8s.  you can try it out really early builds at https://github.com/Azure/aks-engine/blob/96c2dd4d91f2cc7fa807da7c83170267404044a7/docs/topics/features.md#hyper-v-support

marosset commented on content/en/docs/setup/production-environment/windows/intro-windows-in-kubernetes.md:
```diff
@@ -2,118 +2,77 @@
 
-Microsoft maintains a Windows pause infrastructure container at `mcr.microsoft.com/oss/kubernetes/pause:1.4.1`.
+See [Windows OS version support](#windows-os-version-support) for details
+on the supported Windows versions.
```
Or possibly `validated`?
We do run e2e tests that cover everything in that matrix.

sftim commented:
Will rebase shortly (at least, I hope to).

sftim commented:
Rebased with more edits

sftim commented:
Likely to need a rebase if #27585 merges.

kbhawkey commented on content/en/docs/setup/production-environment/windows/intro-windows-in-kubernetes.md:
```diff
@@ -221,222 +203,241 @@ On Windows, you can use the following settings to configure Services and load ba
     1. As native Windows Services
 
-        Kubelet & kube-proxy can be run as native Windows Services using `sc.exe`.
+        You can run the kubelet & kube-proxy as native Windows Services using `sc.exe`.
```
nit: s/&/and/

sftim replied:
```suggestion
        You can run the kubelet and kube-proxy as native Windows Services using `sc.exe`.
```

kbhawkey commented:
Hi. Are there remaining technical questions?

kbhawkey commented on content/en/docs/setup/production-environment/windows/intro-windows-in-kubernetes.md:
```diff
@@ -190,25 +155,42 @@ As outlined above, the [Flannel](https://github.com/coreos/flannel) CNI [meta pl
 
-##### IP address management (IPAM) {#ipam}
+* Windows reference network plugins win-bridge and win-overlay do not implement [CNI spec](https://github.com/containernetworking/cni/blob/master/SPEC.md) v0.4.0, due to a missing "CHECK" implementation.
+* The Flannel VXLAN CNI plugin has the following limitations on Windows:
```
@sftim, Would you double check this list, starting at line 170?
See ,https://deploy-preview-26848--kubernetes-io-master-staging.netlify.app/docs/setup/production-environment/windows/intro-windows-in-kubernetes/#cni-plugin-limitations

sftim commented:
I'll work on a rebase some time, it might not be immediately.

sftim commented:
_rebased_

kbhawkey commented on content/en/docs/setup/production-environment/windows/intro-windows-in-kubernetes.md:
```diff
@@ -4,161 +4,115 @@ reviewers:
 To enable the orchestration of Windows containers in Kubernetes, include Windows nodes in your existing Linux cluster. Scheduling Windows containers in {{< glossary_tooltip text="Pods" term_id="pod" >}} on Kubernetes is similar to scheduling Linux-based containers.
 
-In order to run Windows containers, your Kubernetes cluster must include multiple operating systems, with control plane nodes running Linux and workers running either Windows or Linux depending on your workload needs. Windows Server 2019 is the only Windows operating system supported, enabling [Kubernetes Node](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/arch ...
+In order to run Windows containers, your Kubernetes cluster must include multiple operating systems. The {{< glossary_tooltip text="control plane" term_id="control-plane" >}} is supported on Linux only; however, you can deploy nodes running either Windows or Linux depending on your workload needs. Windows {{< glossary_tooltip text="nodes" term_id="node" >}} are i[supported](#windows-os-version-su ...
```
nit: remove extra "i"

sftim replied:
Thanks!

kbhawkey commented:
These changes need a technical LGTM before merging.
Again, /sig windows

jsturtevant commented on content/en/docs/setup/production-environment/windows/intro-windows-in-kubernetes.md:
```diff
@@ -222,222 +206,240 @@ On Windows, you can use the following settings to configure Services and load ba
 #### IPv4/IPv6 dual-stack
 
-You can enable IPv4/IPv6 dual-stack networking for `l2bridge` networks using the `IPv6DualStack` [feature gate](/docs/reference/command-line-tools-reference/feature-gates/). See [enable IPv4/IPv6 dual stack](/docs/concepts/services-networking/dual-stack#enable-ipv4ipv6-dual-stack) for more details.
+Kubernetes on Windows does not support single-stack "IPv6-only" networking. However,dual-stack IPv4/IPv6 networking for pods and nodes with single-family services is supported.
```
@daschott is single stack IPv6 supported in 2004+?

jsturtevant commented on content/en/docs/setup/production-environment/windows/intro-windows-in-kubernetes.md:
```diff
@@ -222,222 +206,240 @@ On Windows, you can use the following settings to configure Services and load ba
+* `V1.Container.SecurityContext.runAsNonRoot` - Windows does not have a root user. The closest equivalent is ContainerAdministrator which is an identity that doesn't exist on the node.
+* `V1.Container.SecurityContext.runAsUser` - not possible on Windows, no UID support as int.
+* `V1.Container.SecurityContext.seLinuxOptions` - not possible on Windows, no SELinux
+* `V1.Container.terminationMessagePath` - this has some limitations in that Windows doesn't support mapping single files. The default value is `/dev/termination-log`, which does work because it does not exist on Windows by default.
```
since we now have single file mapping, I believe this should be available in containerd deployments 1.21+

sftim replied:
Can you suggest the replacement text @jsturtevant? I don't think I can work out what this should say instead.

jsturtevant replied:
@claudiubelu do you know if this `termination-log` works now with single file mapping containerd?

claudiubelu replied:
Yep, it does. We've also added support for this in Kubernetes v1.17: https://github.com/kubernetes/kubernetes/pull/83057

jsturtevant commented on content/en/docs/setup/production-environment/windows/intro-windows-in-kubernetes.md:
```diff
@@ -222,222 +206,240 @@ On Windows, you can use the following settings to configure Services and load ba
+* `V1.Pod.dnsPolicy` - ClusterFirstWithHostNet - is not supported because Host Networking is not supported on Windows.
+* `V1.Pod.podSecurityContext` - see V1.PodSecurityContext below
+* `V1.Pod.shareProcessNamespace` - this is a beta feature, and depends on Linux namespaces which are not implemented on Windows. Windows cannot share process namespaces or the container's root filesystem. Only the network can be shared.
+* `V1.Pod.terminationGracePeriodSeconds` - this is not fully implemented in Docker on Windows, see: [reference](https://github.com/moby/moby/issues/25982). The behavior today is that the ENTRYPOINT process is sent CTRL_SHUTDOWN_EVENT, then Windows waits 5 seconds by default, and finally shuts down all processes using the normal Windows shutdown behavior. The 5 second default is actually in the Wi ...
```
To be clear we may want to say that it is supported in containerd.

sftim replied:
So `terminationGracePeriodSeconds` works with containerd fully, and partially if your container runtime is Docker. Have I got that right?

jsturtevant replied:
yes

jsturtevant commented on content/en/docs/setup/production-environment/windows/intro-windows-in-kubernetes.md:
```diff
@@ -222,222 +206,240 @@ On Windows, you can use the following settings to configure Services and load ba
+In a Kubernetes Pod, an infrastructure or “pause” container is first created to host the container. In Linux, the cgroups and namespaces that make up a pod need a process to maintain their continued existence; the pause process provides this. Containers that belong to the same pod, including infrastructure and worker containers, share a common network endpoint (same IPv4 and / or IPv6 address, sa ...
 
-Kubernetes on Windows does not support single-stack "IPv6-only" networking. However,dual-stack IPv4/IPv6 networking for pods and nodes with single-family services is supported. See [IPv4/IPv6 dual-stack networking](#ipv4ipv6-dual-stack) for more details.
+Microsoft maintains a Windows pause infrastructure container at `mcr.microsoft.com/oss/kubernetes/pause:1.4.1`. For more details, see the [DOCKERFILE](https://github.com/kubernetes-sigs/windows-testing/blob/master/images/pause/Dockerfile).
```
```suggestion
Microsoft maintains a Windows pause infrastructure container at `mcr.microsoft.com/oss/kubernetes/pause:3.4.1`. For more details, see the [DOCKERFILE](https://github.com/kubernetes-sigs/windows-testing/blob/master/images/pause/Dockerfile).
```

jsturtevant replied:
The pause image is now built along side the same versioning as Linux

sftim replied:
Is it reasonable to remove any mention of a `pause` container image from this page?

jsturtevant replied:
I think we should keep it for a release or two since it was a concern to folks prior to this update

jsturtevant commented on content/en/docs/setup/production-environment/windows/intro-windows-in-kubernetes.md:
```diff
@@ -222,222 +206,240 @@ On Windows, you can use the following settings to configure Services and load ba
 
-#### API
+{{< note >}}
+There is a [known limitation](/docs/tasks/configure-pod-container/configure-gmsa/#gmsa-limitations) when using GMSA with containerd to access Windows network shares, which requires a kernel patch.
```
This has been resolved with patch March 2021+(KB5000822) with the following setting:

```
reg add "HKLM\SYSTEM\CurrentControlSet\Services\hns\State" /v EnableCompartmentNamespace /t REG_DWORD /d 1.
```

ref https://github.com/microsoft/Windows-Containers/issues/44#issuecomment-802404294

jsturtevant commented on content/en/docs/setup/production-environment/windows/intro-windows-in-kubernetes.md:
```diff
@@ -222,222 +206,240 @@ On Windows, you can use the following settings to configure Services and load ba
+| *Kubernetes v1.17* | Windows Server 2019 | Windows Server ver 1809 |
+| *Kubernetes v1.18* | Windows Server 2019 | Windows Server ver 1809, Windows Server ver 1903, Windows Server ver 1909 |
+| *Kubernetes v1.19* | Windows Server 2019 | Windows Server ver 1909, Windows Server ver 2004 |
+| *Kubernetes v1.20* | Windows Server 2019 | Windows Server ver 1909, Windows Server ver 2004 |
```
Is this maintained somewhere else? I believe we recently updated this list for v1.21

sftim replied:
I might have missed that in a rebase - let me check.

sftim replied:
If there's documentation we can link to, we can use that. I did previously consider moving this list of supported versions into the Reference section of the docs, and then linking there.

jsturtevant replied:
the updates for this were merged yesterday: https://github.com/kubernetes/website/pull/26486

jsturtevant commented on content/en/docs/setup/production-environment/windows/intro-windows-in-kubernetes.md:
```diff
@@ -222,222 +206,240 @@ On Windows, you can use the following settings to configure Services and load ba
+1. Apply volume-level encryption using [BitLocker](https://docs.microsoft.com/en-us/windows/security/information-protection/bitlocker/bitlocker-how-to-deploy-on-windows-server).
 
-##### V1.PodSecurityContext
+[RunAsUsername](/docs/tasks/configure-pod-container/configure-runasusername) can be specified for Windows Pods or containers to execute the container processes as a node-default user. This is roughly equivalent to [RunAsUser](/docs/concepts/policy/pod-security-policy/#users-and-groups).
```
We should add that the recommendation would be to run pods with `ContainerUser`.

jsturtevant replied:
or create there own user that is not `ContainerAdministrator`

sftim replied:
I'm not able to infer what to recommend - it's been a few years since I've used Windows, and not at all with Kubernetes, so I don't feel able to explain what readers should do. @jsturtevant would you be willing to suggest the new wording?

jsturtevant replied:
We could add a sentence something like: 

> There are two default accounts for Windows containers: `ContainerAdministrator` and `ContainerUser`.  It is recommended to use `ContainerUser` for most workloads, see more details at [Microsoft documentation on container user accounts](https://docs.microsoft.com/en-us/virtualization/windowscontainers/manage-containers/container-security#when-to-use-containeradmin-and-containeruser-user-accounts)

jsturtevant commented on content/en/docs/setup/production-environment/windows/intro-windows-in-kubernetes.md:
```diff
@@ -459,165 +461,163 @@ Your main source of help for troubleshooting your Kubernetes cluster should star
+1. My Pods are stuck at "Container Creating" or restarting over and over
 
-        For initial troubleshooting, you can use the following flags in [nssm.exe](https://nssm.cc/) to redirect stdout and stderr to a output file:
+   Check that your pause image is compatible with your OS version. The [instructions](https://docs.microsoft.com/en-us/virtualization/windowscontainers/kubernetes/deploying-resources) assume that both the OS and the containers are version 1803. If you have a later version of Windows, such as an Insider build, you need to adjust the images accordingly.
```
```suggestion
   Check that your pause image is compatible with your OS version. The [instructions](https://docs.microsoft.com/en-us/virtualization/windowscontainers/kubernetes/deploying-resources) assume that both the OS and the containers are the same version (i.e 2019 LTSC). If you have a later version of Windows, such as an Insider build, you need to adjust the images accordingly.
```

jsturtevant commented on content/en/docs/setup/production-environment/windows/intro-windows-in-kubernetes.md:
```diff
@@ -459,165 +461,163 @@ Your main source of help for troubleshooting your Kubernetes cluster should star
 
-    This was implemented in Kubernetes 1.15 by including wincat.exe in the pause infrastructure container `mcr.microsoft.com/oss/kubernetes/pause:1.4.1`. Be sure to use these versions or newer ones.
-    If you would like to build your own pause infrastructure container be sure to include [wincat](https://github.com/kubernetes-sigs/sig-windows-tools/tree/master/cmd/wincat).
+   This was implemented in Kubernetes 1.15 by including `wincat.exe` in the pause infrastructure container `mcr.microsoft.com/oss/kubernetes/pause:1.4.1`. Be sure to use a supported version of Kubernetes.
```
```suggestion
   This was implemented in Kubernetes 1.15 by including `wincat.exe` in the pause infrastructure container `mcr.microsoft.com/oss/kubernetes/pause:3.4.1`. Be sure to use a supported version of Kubernetes.
```

sftim replied:
The original page was incorrect about the version that included the fix?

jsturtevant commented on content/en/docs/setup/production-environment/windows/intro-windows-in-kubernetes.md:
```diff
@@ -459,165 +461,163 @@ Your main source of help for troubleshooting your Kubernetes cluster should star
-    This was implemented in Kubernetes 1.15 by including wincat.exe in the pause infrastructure container `mcr.microsoft.com/oss/kubernetes/pause:1.4.1`. Be sure to use these versions or newer ones.
-    If you would like to build your own pause infrastructure container be sure to include [wincat](https://github.com/kubernetes-sigs/sig-windows-tools/tree/master/cmd/wincat).
+   This was implemented in Kubernetes 1.15 by including `wincat.exe` in the pause infrastructure container `mcr.microsoft.com/oss/kubernetes/pause:1.4.1`. Be sure to use a supported version of Kubernetes.
+   If you would like to build your own pause infrastructure container be sure to include [wincat](https://github.com/kubernetes-sigs/sig-windows-tools/tree/master/cmd/wincat).
```
```suggestion
   If you would like to build your own pause infrastructure container be sure to include [wincat](https://github.com/kubernetes/kubernetes/tree/master/build/pause/windows/wincat).
```

jsturtevant commented on content/en/docs/setup/production-environment/windows/intro-windows-in-kubernetes.md:
```diff
@@ -639,25 +639,18 @@ If filing a bug, please include detailed information about how to reproduce the
+For a detailed explanation of Windows distribution channels see the [Microsoft documentation](https://docs.microsoft.com/en-us/windows-server/get-started-19/servicing-channels-19).
 
-### Deployment with kubeadm and cluster API
+Information on the different Windows Server servicing channels including their support models can be found at [Windows Server servicing channels](https://docs.microsoft.com/en-us/windows-server/get-started-19/servicing-channels-19).
```
This line and the distribution channels line right above it link to the same page.  I believe "distribution channels" and "servicing channels" mean the same thing in this context and only need to be referenced once.  @immuzz could you confirm?

sftim commented:
I'll rebase this soon, incorporating both feedback and the changes made since this PR's current merge base.

sftim commented:
(for some value of “soon”)

sftim commented:
This lays some groundwork for working on #29769

kbhawkey commented:
👋 If there are no further comments, I suggest to move this PR forward.
Ping @kubernetes/sig-windows-leads for technical review

sftim commented on content/en/docs/setup/production-environment/windows/intro-windows-in-kubernetes.md:
```diff
@@ -1,1335 +1,968 @@
+for network, process and file system isolation. The Linux cgroup APIs can be used
+to gather CPU, I/O, and memory use statistics.
+
+In contrast, Windows uses a _job object_ per container with a system namespace filter
```
I'm not sure this is right. Is the “job object” obviously within the container?

sftim commented on content/en/docs/setup/production-environment/windows/intro-windows-in-kubernetes.md:
```diff
@@ -1,1335 +1,968 @@
+for memory usage on the node (outside of containers), and reduce
+[NodeAllocatable](/docs/tasks/administer-cluster/reserve-compute-resources/#node-allocatable)).
+As you deploy workloads, set resource limits on containers. This also subtracts from
+`NodeAllocatable` and prevents the scheduler from adding more pods once a node is full.
```
We typically wouldn't write pods in title case here. Other localizations might, especially where “pods” is not a word in the target language.

sftim commented on content/en/docs/setup/production-environment/windows/intro-windows-in-kubernetes.md:
```diff
@@ -1,1335 +1,968 @@
 workloads on Kubernetes. However, on their own they are not enough to enable
 the proper lifecycle management of Windows workloads in a dynamic cloud native
-environment. We added support for the following features:
+environment. Kubernetes also supports:
```
I'm not sure what the original text meant to imply!

marosset replied:
I think maybe these functionalities were not available when Windows container support went stable (v1.14) but has since been added?

I wonder if we should just remove parts of this section because I am also not sure what the original text meant to imply.

sftim commented on content/en/docs/setup/production-environment/windows/intro-windows-in-kubernetes.md:
```diff
@@ -1,1335 +1,968 @@
-* V1.Pod.volumeDevices - this is a beta feature, and is not implemented on
-  Windows. Windows cannot attach raw block devices to pods.
+   The kubelet and kube-proxy are already configured to run as native Windows Services,
+   offering resiliency by re-starting the services automatically in the event of
```
These are not Services in the Kubernetes API. “services” is less likely to mislead.

sftim commented on content/en/docs/setup/production-environment/windows/intro-windows-in-kubernetes.md:
```diff
@@ -1,1335 +1,968 @@
+   The kubelet and kube-proxy are already configured to run as native Windows Services,
+   offering resiliency by re-starting the services automatically in the event of
+   failure (for example a process crash). You have two options for configuring these
+   node components as services.
```
These are not Services in the Kubernetes API. “services” is less likely to mislead.

sftim commented:
My understanding is that there's unaddressed feedback and that, as things stand, these are all nits.

If so, I agree: we can try to move this forward towards merge. Any remaining nits could turn into small issues, perhaps good-first-issues.

jsturtevant commented on content/en/docs/setup/production-environment/windows/intro-windows-in-kubernetes.md:
```diff
@@ -1,1335 +1,968 @@
+more CPU time slices when compared to other processes running on the Windows host.
+More information on the allowable values and their meaning is available at
+[Windows Priority Classes](https://docs.microsoft.com/en-us/windows/win32/procthread/scheduling-priorities#priority-class).
+To ensure that running Pods do not starve the kubelet of CPU cycles, set this flag to `ABOVE_NORMAL_PRIORITY_CLASS` or above.
```
these are constants defined here: https://docs.microsoft.com/en-us/windows/win32/procthread/scheduling-priorities#priority-class (link is reference in the paragraph above.  You have to choose one of these constants otherwise kubelet fails to start: https://github.com/kubernetes/kubernetes/blob/c1e69551be1a72f0f8db6778f20658199d3a686d/cmd/kubelet/app/init_windows.go#L56-L60

kbhawkey commented:
Thanks @sftim .
At some point, this page could be split into several pages.
/lgtm

marosset commented on content/en/docs/setup/production-environment/windows/intro-windows-in-kubernetes.md:
```diff
@@ -1,1335 +1,965 @@
+(Job objects are a Windows process isolation mechanism and are different from
+what Kubernetes refers to as a {{< glossary_tooltip term_id="job" text="Job" >}}).
+
+There is no way to run a Windows container without the namespace filtering in
```
We have 'HostProcess' containers now (alpha in v1.22 and hopefully beta in v1.23) which are similar to privilege containers and can access / manipulate host resources.
We can update this section with a follow on PR tho :)

sftim replied:
Please do! Or, even better, split that off into a separate page. Remember, this page is _supposed_ to be about setting up a cluster.

marosset replied:
We have https://kubernetes.io/docs/tasks/configure-pod-container/create-hostprocess-pod/
I was going to update that info to point to that page :)

marosset commented on content/en/docs/setup/production-environment/windows/intro-windows-in-kubernetes.md:
```diff
@@ -1,1335 +1,965 @@
+{{< /note >}}
+
+On Windows, good practice to avoid over-provisioning is to configure the kubelet
+with a system reserved memory of at least 2GiB to account for Windows, Kubernetes
```
```suggestion
with a system reserved memory of at least 2GiB to account for Windows, Kubernetes,
```

marosset commented on content/en/docs/setup/production-environment/windows/intro-windows-in-kubernetes.md:
```diff
@@ -1,1335 +1,965 @@
+
+* HugePages: not supported for Windows containers
+* Privileged containers: not supported for Windows containers
+* TerminationGracePeriod: not implemented
```
```suggestion
* TerminationGracePeriod: required containerD
```

marosset commented on content/en/docs/setup/production-environment/windows/intro-windows-in-kubernetes.md:
```diff
@@ -1,1335 +1,965 @@
+  POSIX systems such as Linux use a bitmask based on object permissions and UID+GID,
+  plus _optional_ access control lists.
+* File paths - the convention on Windows is to use `\` instead of `/`. The Go IO
+  libraries typically accept both and just make it work, but when you're setting a
```
```suggestion
  libraries and PowerShell typically accept both and just make it work, but when you're setting a
```

marosset commented:
/lgtm
I left a few minor suggestions but overall I think this is ready to merge.
Thank you so much @sftim for helping out here!

kbhawkey commented:
Page preview:
https://deploy-preview-26848--kubernetes-io-main-staging.netlify.app/docs/setup/production-environment/windows/intro-windows-in-kubernetes/

Thanks @sftim 
Great effort everyone. Thanks for the reviews and comments!
/approve 🎉
